datafusion/core/src/optimizer/subquery_filter_to_join.rs

```diff
@@ -0,0 +1,334 @@
+            ))?
+            .project(vec![col("test.b")])?
+            .build()?;
+
+        let expected = "Projection: #test.b [b:UInt32]\
+        \n  Semi Join: #test.b = #test.c [a:UInt32, b:UInt32, c:UInt32]\
+        \n    Semi Join: #test.c = #test.c [a:UInt32, b:UInt32, c:UInt32]\
+        \n      TableScan: test projection=None [a:UInt32, b:UInt32, c:UInt32]\
+        \n      Projection: #test.c [c:UInt32]\
+        \n        TableScan: test projection=None [a:UInt32, b:UInt32, c:UInt32]\
+        \n    Projection: #test.c [c:UInt32]\
+        \n      TableScan: test projection=None [a:UInt32, b:UInt32, c:UInt32]";
+
+        assert_optimized_plan_eq(&plan, expected);
+        Ok(())
+    }
+
+    /// Test for IN subquery with additional filters
+    #[test]
+    fn in_subquery_with_filters() -> Result<()> {
+        let table_scan = test_table_scan()?;
+        let plan = LogicalPlanBuilder::from(table_scan)
+            .filter(and(
+                in_subquery(col("c"), test_subquery()?),
+                and(
+                    binary_expr(col("a"), Operator::Eq, lit(1_u32)),
+                    binary_expr(col("b"), Operator::Lt, lit(30_u32)),
+                ),
+            ))?
+            .project(vec![col("test.b")])?
+            .build()?;
+
+        let expected = "Projection: #test.b [b:UInt32]\
+        \n  Filter: #test.a = UInt32(1) AND #test.b < UInt32(30) [a:UInt32, b:UInt32, c:UInt32]\
+        \n    Semi Join: #test.c = #test.c [a:UInt32, b:UInt32, c:UInt32]\
+        \n      TableScan: test projection=None [a:UInt32, b:UInt32, c:UInt32]\
+        \n      Projection: #test.c [c:UInt32]\
+        \n        TableScan: test projection=None [a:UInt32, b:UInt32, c:UInt32]";
+
+        assert_optimized_plan_eq(&plan, expected);
+        Ok(())
+    }
+
+    /// Test for nested IN subqueries
+    #[test]
+    fn in_subquery_nested() -> Result<()> {
+        let table_scan = test_table_scan()?;
+
+        let subquery = LogicalPlanBuilder::from(table_scan.clone())
+            .filter(in_subquery(col("a"), test_subquery()?))?
+            .project(vec![col("a")])?
+            .build()?;
+
+        let plan = LogicalPlanBuilder::from(table_scan)
+            .filter(in_subquery(col("b"), Arc::new(subquery)))?
+            .project(vec![col("test.b")])?
+            .build()?;
+
+        let expected = "Projection: #test.b [b:UInt32]\
+        \n  Semi Join: #test.b = #test.a [a:UInt32, b:UInt32, c:UInt32]\
+        \n    TableScan: test projection=None [a:UInt32, b:UInt32, c:UInt32]\
+        \n    Projection: #test.a [a:UInt32]\
+        \n      Semi Join: #test.a = #test.c [a:UInt32, b:UInt32, c:UInt32]\
+        \n        TableScan: test projection=None [a:UInt32, b:UInt32, c:UInt32]\
+        \n        Projection: #test.c [c:UInt32]\
+        \n          TableScan: test projection=None [a:UInt32, b:UInt32, c:UInt32]";
+
+        assert_optimized_plan_eq(&plan, expected);
+        Ok(())
+    }
+}
```

Comment:
I think it would also be helpful to coverage of the negative cases (aka cases that can't be rewritten like `x IN (select ...) OR y = 5`?

Reply:
Added cases for most complex expression and for filter input being rewritten while filter remains untouched (check if falling back to original query doesn't affect its recursive call results)